### PR TITLE
Bhcal threshold

### PR DESCRIPTION
**Pull request by kkauder**
### Briefly, what does this PR introduce?
Updated BHCAL thresholds and digitization parameters from the [digitization table from 10/2](https://brookhavenlab.sharepoint.com/:x:/s/EICPublicSharingDocs/EeICxMJ9vGtPtyq5g4qzcSwBKCqkHDj06KDdSvYQ7pxqrQ?rtime=lD_5kMnT20g).

### Notes:
- This uses a cut on accumulated energy. Instead, you could set the energy cut to 0 and use `.thresholdValue=328` instead (or `.thresholdFactor = 164`). The difference between the two is that in the latter case, an ADC fluctuation can bring the result below/above threshold.
- I believe the 5 MeV threshold corresponds to a 150 MeV real deposition, accounting for sampling fraction. Just double-checking that that's correct.

### What kind of change does this PR introduce?
- [x] Bug fix (issue #895 )
- [ ] New feature (issue #__)
- [ ] Documentation update
- [x] Other: Maintenance

### Please check if this PR fulfills the following:
- [ ] Tests for the changes have been added
- [ ] Documentation has been added / updated
- [x] Changes have been communicated to collaborators

### Does this PR introduce breaking changes? What changes might users need to make to their code?
No.

### Does this PR change default behavior?
Yes.

### CODE REVIEW

**Comment by johnlajoie** [comment]
I've sked Derek to take a look since he is the one who has worked on the BHCal digitization.  Since we read out individual tiles, the sampling has no meaning in this context (it's not a tower or some such). The 5 MeV is directly on the amount of energy deposited in the tile, which corresponds to about a half-MIP or ~20 SiPM pixels.

**Comment by kkauder** [comment]
> I've asked Derek to take a look since he is the one who has worked on the BHCal digitization. Since we read out individual tiles, the sampling has no meaning in this context (it's not a tower or some such). The 5 MeV is directly on the amount of energy deposited in the tile, which corresponds to about a half-MIP or ~20 SiPM pixels.

What is the sampling fraction used for then?

If I follow the digitization here
https://github.com/eic/EICrecon/blob/main/src/algorithms/calorimetry/CalorimeterHitDigi.cc#L171-L172
I get for a 5 MeV deposition
`adc = round(300 + 5 MeV * (1.0 / 1000 MeV) * 65536) = 628.`
Plugging that into 
https://github.com/eic/EICrecon/blob/main/src/algorithms/calorimetry/CalorimeterHitReco.cc#L171-L172
I get 
`energy = (628 - 300 ) / 65536 * 1000 MeV / (sampfrac=0.033) = 152 MeV`
In other words, the digi class cuts on and converts the energy as deposited, the reco class corrects that by the sampling fraction.

File: src/detectors/BHCAL/BHCAL.cc [hunk 1]
```diff
@@ -42,13 +49,13 @@ extern "C" {
...
-            .dyRangeADC = 1.0 * dd4hep::GeV,
-            .pedMeanADC = 10,
-            .pedSigmaADC = 2.0,
-            .resolutionTDC = 1.0 * dd4hep::picosecond,
-            .thresholdFactor = 5.0,
-            .thresholdValue = 1.0,
+            .capADC        = HcalBarrel_capADC,
+            .dyRangeADC    = HcalBarrel_dyRangeADC,
+            .pedMeanADC    = HcalBarrel_pedMeanADC,
+            .pedSigmaADC   = HcalBarrel_pedSigmaADC, // not used; relying on energy cut
```
Comment (ruse-traveler) [inline src/detectors/BHCAL/BHCAL.cc:55]: Apologies, but I wasn't able to make the background meeting yesterday. Was there a decision to use the energy threshold vs. the ADC threshold?

Since we'll cut on ADC in actual data, I think it would be better to move to relying on only the ADC threshold.

**Review by ruse-traveler — COMMENTED** [review]
This all looks quite reasonable to me!

Like I mentioned in my comments above, I think it would be preferable to move to relying on only the ADC threshold, and the math you suggest above looks good.

Truth be told, though, I'm having a hard time telling if the changes I see in the artifacts are meaningful. So let me run a quick sanity check to make sure that using the above numbers without the energy threshold produce the same reconstructed hit spectrum.

File: src/detectors/BHCAL/BHCAL.cc (hunk 1 above)
Comment (kkauder) [reply src/detectors/BHCAL/BHCAL.cc:55]: There hasn't been a decision, just to go through the calo ones individually and where necessary/easy default to the energy cut. If you prefer the ADC cut, just give me the okay (and add the number to the spreadsheet), and I will make the change.

Comment (ruse-traveler) [reply src/detectors/BHCAL/BHCAL.cc:55]: Ah! Gotcha! In that case, let's plan on going to ADC only. Let me double check that the 2 are indeed equivalent, and then we can make the changes accordingly...

Comment (ruse-traveler) [reply src/detectors/BHCAL/BHCAL.cc:55]: Hi Kolja, I finished up running some checks. I've attached a plot comparing the reconstructed hit spectrum for different combinations of energy cuts/adc thresholds (made w/ 100 10 GeV single pi-).

I should have caught this earlier, but we do want to cut on a half-MIP (5 MeV) deposition without any adjustment for sampling fractions (like John noted). So I think I may be misunderstanding what these parameters are...

We've set the pedMeanADC to be 300 ADC, just under 5 MeV. Following the logic in the digi factory, the condition for zero-suppression is:

`amplitude < (threshold + pedMeanADC)`

which means that if we want to cut anything below 300 (333) ADC then we would need to set the threshold parameters to something like

```
thresholdFactor = 0 (1)
thresholdValue = 0 (31)
```

In the plot, the yellow line corresponds to to a zero-suppression of 300 (factor/value = 0) while the green line corresponds to a zero-suppression of 333 (factor = 1, value = 31).

So I was wondering: do `thresholdFactor` and `thresholdValue` correspond to physical quantities? As I understand it, `thresholdFactor` corresponds to an n-sigma cut on our pedestal distribution, but I'm confused about what `thresholdValue` corresponds to...

checkingCuts_withVal31fac1AndVal0Fac0_PR1095 bhcalRecHitsEne d3m11y2023

**Comment by kkauder** [comment]
I put together a hopefully better explanation for yesterday's Simulation meeting, https://indico.bnl.gov/event/21036/
Basically, the important line is 
```
thresholdADC = m_cfg.thresholdFactor * m_cfg.pedSigmaADC + m_cfg.thresholdValue;
```
You can cut EITHER by specifying the number of SigmaADC's OR a fixed threshold value, one of the two should always be 0. So, you can set `thresholdFactor=3;  pedSigmaADC = 31; thresholdValue = 0;` or `thresholdFactor=0;  pedSigmaADC = 0; thresholdValue = 33;` to get to 333. 
It's all a bit redundant, but it's grown over time.

**Comment by ruse-traveler** [comment]
> I put together a hopefully better explanation for yesterday's Simulation meeting, https://indico.bnl.gov/event/21036/ Basically, the important line is

Perfect! Thanks! That helps quite a lot!

> You can cut EITHER by specifying the number of SigmaADC's OR a fixed threshold value, one of the two should always be 0. So, you can set `thresholdFactor=3; pedSigmaADC = 31; thresholdValue = 0;` or `thresholdFactor=0; pedSigmaADC = [whatever is right, only used in digi]; thresholdValue = 33;` to get to 333. It's all a bit redundant, but it's grown over time.
 
Gotcha! In that case, let's go with the latter case (`pedSigmaADC = 300, thresholdFactor = 0, thresholdValue = 33`). Conceptually, that matches most closely with what we've been doing! I'll put the suggestions in the code...

File: src/detectors/BHCAL/BHCAL.cc [hunk 2]
```diff
@@ -42,13 +49,13 @@ extern "C" {
...
+            .resolutionTDC = HcalBarrel_resolutionTDC,
+            .thresholdFactor = 0.0, // not used; relying on energy cut
+            .thresholdValue = 0.0, // not used; relying on energy cut
```
Comment (ruse-traveler) [inline src/detectors/BHCAL/BHCAL.cc:58]: Suggestion:
```
.thresholdValue = 33.0, // pedSigmaADC + thresholdValue = half-MIP (333 ADC)

**Review by ruse-traveler — CHANGES_REQUESTED** [review]

File: src/detectors/BHCAL/BHCAL.cc [hunk 3]
```diff
@@ -22,18 +22,25 @@ extern "C" {
...
+        decltype(CalorimeterHitDigiConfig::pedMeanADC)    HcalBarrel_pedMeanADC = 300;
+        decltype(CalorimeterHitDigiConfig::pedSigmaADC)   HcalBarrel_pedSigmaADC = 2;
+        decltype(CalorimeterHitDigiConfig::resolutionTDC) HcalBarrel_resolutionTDC = 1 * dd4hep::picosecond;
+
         app->Add(new JChainMultifactoryGeneratorT<CalorimeterHitDigi_factoryT>(
           "HcalBarrelRawHits", {"HcalBarrelHits"}, {"HcalBarrelRawHits"},
           {
             .eRes = {},
             .tRes = 0.0 * dd4hep::ns,
             .threshold = 5.0 * dd4hep::MeV,
```
[the remark this answers is not in the document]
Comment (kkauder) [reply src/detectors/BHCAL/BHCAL.cc]: These values are still needed for digitization. Everything else, I believe I implemented, but please double-check.

[the remark this answers is not in the document]
Comment (ruse-traveler) [reply src/detectors/BHCAL/BHCAL.cc]: Whoops! I accidentally commented on the wrong lines 😅 

Looks great!

**Review by ruse-traveler — APPROVED** [review]
